plugins/permission-common/src/PermissionClient.ts

```diff
@@ -41,7 +41,7 @@ const permissionCriteriaSchema: z.ZodSchema<
     .object({
       rule: z.string(),
       resourceType: z.string(),
-      params: z.record(z.any()),
+      params: z.record(z.any()).optional(),
```

Comment:
@kuangp I'm not sure this is a bug linked (or a bug at all). Going back to before these changes the original line was ...

```ts
params: z.array(z.unknown()),
```

... without optional. And checking out a previous commit, the same test added above fails. Are you seeing a change in functionality somewhere, because if you are, and I'm correct with the above, it might be the bug exists elsewhere.

Reply:
@HHogg digging into this, here's what I suspect caused the bug here:

[Originally](https://github.com/backstage/backstage/pull/13905/files#diff-86bf8b2c2693596aa362974dcec18329152d803e6ec1b2ab8cbb8caf2c29c616L40) `createConditionFactory` was grabbing `params` via  [rest parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters). For rules that did not receive any params, this will always still result in an empty array due to the semantics of rest params.  This allowed the schema validation of `z.array(z.unknown())` to always pass before since the `params` field was always defined even if empty. This is also why the newly added test fails in the previous implementation since that is not a scenario that could actually arise so that test is not applicable before. This was not technically a bug before but is a new bug that came out of the recent changes because of changed behavior.

With the change from your PR, `params` became potentially `undefined` for rules that don't take any params and results in the schema validation `ZodError` reported in #14228 not matching `z.record(z.any())`. 

I think this change was just mistakenly left out from your PR since I see a similar schema update in [createPermissionIntegrationRouter](https://github.com/backstage/backstage/pull/13905/files#diff-7a65a8dbe156fbea873fa051cedb9605c5eb98bb19965b6d11acf04e5ba5fd31R50)

Let me know if I misunderstood anything here!